File: Game/src/main/java/net/winternetwork/bedwars/game/module/map/object/GameMap.java
package net.winternetwork.bedwars.game.module.map.object;

import lombok.Builder;
import lombok.Getter;
import lombok.ToString;
import net.winternetwork.bedwars.api.game.map.IGameMap;
import net.winternetwork.bedwars.game.module.map.object.adapter.LocationAdapter;
import org.bukkit.Location;
import org.bukkit.configuration.ConfigurationSection;

@Getter
@Builder
@ToString
public class GameMap implements IGameMap {

    private final String name;

    private final int maxPlayers;

    private final Location lobbyLocation;

    private final Location[] spawnLocations;

    public static GameMap fromSection(ConfigurationSection section) {
        LocationAdapter locationAdapter = new LocationAdapter();

        Location[] spawnLocations = new Location[section.getInt("maxPlayers")];

        for (String id : section.getConfigurationSection("locations").getKeys(false)) {
            int i = Integer.parseInt(id);

            spawnLocations[i - 1] = locationAdapter.from(
                    section.getString("locations." + i)
            );
        }

        return GameMap.builder()
                .name(section.getName())
                .maxPlayers(section.getInt("maxPlayers"))
                .lobbyLocation(
                        locationAdapter.from(
                                section.getString("lobbyLocation")
                        )
                )
                .spawnLocations(spawnLocations)
                .build();
    }

    public void setLocation(int id, Location location) {
        spawnLocations[id - 1] = location;
    }
}

File: Game/src/main/kotlin/net/winternetwork/bedwars/modules/maps/model/GameMap.kt
package net.winternetwork.bedwars.modules.maps.model

import net.winternetwork.bedwars.api.game.map.IGameMap
import net.winternetwork.bedwars.api.storage.adapter.Adapter
import org.bukkit.Bukkit
import org.bukkit.Location
import org.bukkit.configuration.ConfigurationSection

data class GameMap(
    override val name: String,
    override val maxPlayers: Int,
    override val lobbyLocation: Location,
    override val locations: ArrayList<Location> = arrayListOf()
) : IGameMap {

    fun setLocation(id: Int, location: Location) {
        locations[id - 1] = location
    }

    companion object {
        fun fromSection(section: ConfigurationSection): GameMap {

            val maxPlayers = section.getInt("maxPlayers")
            val locations = arrayListOf<Location>()

            for (id in section.getConfigurationSection("locations").getKeys(false)) {
                val i = id.toInt()

                locations[i - 1] = LocationAdapter.from(
                    section.getString("locations.$i")
                )
            }

            return GameMap(
                section.name,
                maxPlayers,
                LocationAdapter.from(section.getString("lobbyLocation")),
                locations
            )
        }
    }

}

object LocationAdapter : Adapter<Location, String> {

    override fun from(c: String): Location {
        val split = c.split(";")
        return Location(
            Bukkit.getWorld(split[0]),
            split[1].toDouble(),
            split[2].toDouble(),
            split[3].toDouble()
        )
    }

    override fun to(t: Location): String {
        return ArrayList<String>().apply {
            add(t.world.name)
            add(t.x.toString())
            add(t.y.toString())
            add(t.z.toString())
        }.joinToString(";")
    }
}